machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,243 @@
+	// This field is populated from the machine-config-osimageurl configmap in the openshift-machine-config-operator namespace.
+	// It will come in the format: 4.16.0-0.nightly-2024-04-03-065948 or any valid release. The MachineOSBuilder populates this field and validates that this is a valid stream.
+	// This is used as a label in the dockerfile that builds the OS image.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +optional
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// this is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// you can specify up to 7 containerFiles
+	// +patchMergeKey=containerfileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=7
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerfileArch"`
+}
+
+// BuildOutputs holds all information needed to handle booting the image after a build
+// +union
+type BuildOutputs struct {
+	// currentImagePullSecret is the secret used to pull the final produced image.
+	// must live in the openshift-machine-config-operator namespace
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	CurrentImagePullSecret ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=JobImageBuilder
+	// Valid options are: JobImageBuilder
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noarch
```

Comment:
These should be PascalCase normally, and not lowercased like this, the proper name for arm64 is ARM64 anyway, for example

Also, AMD64 and x86_64 are the same thing? AArch64 and ARM64 are the same?

Is noarch a special constant for this implementation or something borrowed from elsewhere?

Reply:
Will check on this point

Reply:
> AMD64 and x86_64 are the same thing? AArch64 and ARM64 are the same?

Yes they are. For some reason, `$ uname -m` returns `x86_64`, but anything Golang-related wants `amd64`. The same applies to AArch64 and ARM64.

> Is noarch a special constant for this implementation?

It is. Basically, if a Containerfile is associated with noarch, it means that the same Containerfile should be able to be built for any supported architecture. There are situations where one may want to supply a different Containerfile for a different architecture. The most common example I can think of off the top of my head is when someone is installing an architecture-specific script or tool and doing something like `curl "https://host.com/$(uname -m)/tool.rpm"` doesn't cut it.

Reply:
> Yes they are. For some reason, $ uname -m returns x86_64, but anything Golang-related wants amd64. The same applies to AArch64 and ARM64.

We should make sure we only have one version for each real architecture (the go one typically), to avoid duplication within the enum. If you need to convert, convert within the operator.

> Basically, if a Containerfile is associated with noarch, it means that the same Containerfile should be able to be built for any supported architecture.

When there are many containerfiles configured, are they supposed to be different architectures? What happens if I were to specify multiple noarch containerfiles in the list?